app/models/mixins/authentication_mixin.rb

```diff
@@ -443,8 +443,14 @@ def assign_nested_endpoint(attributes)
   end
 
   def assign_nested_authentication(attributes)
-    record = authentications.where(:authtype => attributes['authtype']).first_or_initialize
     klass = authentication_class(attributes)
+    record = authentications.where(:authtype => attributes['authtype']).first
+
+    if record.nil?
+      record = klass.new(:resource => self)
+      authentications << record
+    end
```

Comment:
I wanted to do
```
record = authentications.where(:authtype => attributes['authtype']).first
record ||= authentications.build(:type => klass.to_s)
```

But I end up running into:
```
     RuntimeError:
       AuthToken is not a subclass of AuthUseridPassword
     # ./app/models/mixins/new_with_type_sti_mixin.rb:11:in `new'
     # ./app/models/authentication.rb:11:in `new'
     # ./app/models/authentication.rb:9:in `new'
```

Reply:
What about?

```ruby
    klass = authentication_class(attributes)
    record = authentications.where(:authtype => attributes['authtype'], :type => klass).first_or_initialize
```

Reply:
You have the same issue of `record` being an `AuthUseridPassword` type
```
authentications.where(:authtype => attributes['authtype'], :type => klass).first_or_initialize.class.name
"AuthUseridPassword"
```

Reply:
Oh I misunderstood...I thought the problem was on the search side...flipping it them, how about:

```ruby
    klass = authentication_class(attributes)
    record = authentications.where(:authtype => attributes['authtype']).first_or_initialize(:type => klass)
```

Reply:
Then you run into the same issue as `authentications.build(:type => klass.to_s)`:

```
  1) ExtManagementSystem.create_from_params with an auth_token type authentication creates an AuthToken type authentication record
     Failure/Error: raise _("%{class_name} is not a subclass of %{name}") % {:class_name => klass.name, :name => name}
     
     RuntimeError:
       AuthToken is not a subclass of AuthUseridPassword
     # ./app/models/mixins/new_with_type_sti_mixin.rb:11:in `new'
     # ./app/models/authentication.rb:11:in `new'
     # ./app/models/authentication.rb:9:in `new'
     # /home/grare/adam/.gems/2.7.0/gems/activerecord-5.2.4.4/lib/active_record/reflection.rb:154:in `build_association'
     # /home/grare/adam/.gems/2.7.0/gems/activerecord-5.2.4.4/lib/active_record/associations/association.rb:284:in `build_record'
     # /home/grare/adam/.gems/2.7.0/gems/activerecord-5.2.4.4/lib/active_record/associations/collection_association.rb:108:in `build'
     # /home/grare/adam/.gems/2.7.0/gems/activerecord-5.2.4.4/lib/active_record/association_relation.rb:19:in `block in build'
     # /home/grare/adam/.gems/2.7.0/gems/activerecord-5.2.4.4/lib/active_record/relation.rb:281:in `scoping'
     # /home/grare/adam/.gems/2.7.0/gems/activerecord-5.2.4.4/lib/active_record/association_relation.rb:19:in `build'
     # /home/grare/adam/.gems/2.7.0/gems/activerecord-5.2.4.4/lib/active_record/relation.rb:112:in `first_or_initialize'
     # ./app/models/mixins/authentication_mixin.rb:447:in `assign_nested_authentication'
```

Reply:
The only way to get around this I think is to change how we initialize Authentication objects, e.g. if we change this:
```
def self.new(*args, &block)
  if self == Authentication
    AuthUseridPassword.new(*args, &block)
  else
    super
  end 
end
```

To something like this:
```
def self.new(*args, &block)
  if self == Authentication && (h = args.first).kind_of?(Hash)
    h[:type] ||= "AuthUseridPassword"
  end 

  super
end
```

Then we would actually be able to use `:type` the way you would expect to

Reply:
I think I prefer that change because the unexpected nature of the :type not being honored is what's confusing.

Reply:
Only thing that would also need to be added is to check for both `:type` and `"type"`

See https://github.com/ManageIQ/manageiq/blob/master/app/models/mixins/new_with_type_sti_mixin.rb#L8 for probably the "best" way to get at the correct value.

Reply:
Done, I'm going to push a cross-repo test PR since this is more of a fundamental change